Topics/Software_Engineering/XP.md

```diff
@@ -0,0 +1,100 @@
+
+**5. Respect**
+* This involves valuing customer input, even if they are not as technically compotent. This also involves respect amongst developers, for each other's work and also for the product itself. This respectful environment fosters better work.
+
+### Principles
+
+There are a lot of principles when it comes to XP, but here we have summarized ones that would affect developers the most.
+
+**1. Pair programming**
+* Two developers should work together at one computer, promoting collaboration, knowledge sharing and hence producing higher quality code.
+
+**2. Whole team**
+* Not only team members should be in the development process, but also customers and business stakeholders. This spreads the responsibility of the project's success between everyone, and also helps keep the project keep on track during changes.
+
+**3. Continuous integration**
+* Change should be embraced throughout the project, even if not originally decided upon.
+
+**4. Test-first programming**
+* Instead of the typical process of first writing code then writing tests, the path changes to writing tests, then writing code to pass those test. This reduces the amount of bugs.
+
+**5. Short development cycles**
+* Sprints should be done in 1-2 week durations.
+
+**6. 10-minute build**
+* Building the whole system should be done automatically, and all tests should be able to run within 10 minutes. The 10 minute limit is because any longer build times would disencourage frequent builds, which inevitably introduces more errors over time. 
+
+
+## Comparison with other agile methodologies
+
+### XP vs Scrum
+
+**1. Duration of sprints.** Scrum teams usually work on sprints that are 2 weeks to 1 month long. XP sprints are usually only 1 to 2 weeks long.
+
+**2. Flexibility during sprints.** For Scrum teams, once the sprint planning has been done, the expectation is to deliver those items by the end of the deadline. However, in XP, due to the higher emphasis on responsiveness, as long as the team hasn't already started work on a feature, another feature of similar workload can be swapped in.
+
+**3. XP prescribes engineering practices.** As we have discussed, XP has a lot of specific implementation details that developers need to follow.
+
+### XP vs Waterfall
+
+**1. Plan as you go.** Instead of a separate planning phase, planning takes place at the beginning of each sprint, which only lasts 1-2 weeks each.
+
+**2. XP tests before, Waterfall tests after.** In XP, tests are written before code is even implemented.
+
+**3. Smaller chunk-based contributions.** Instead of working on big features alone for a long period of time, developers work on small chunks and push as they go. This avoids massive merge conflicts in the future, saving time.
```

Comment:
You might want links to the respective Scrum and Waterfall pages on the wiki in case the reader in interested in the other methodologies.

Reply:
Good idea